hypex/comparators/abstract.py

```diff
@@ -125,41 +138,168 @@ def _extract_dataset(
+            compare_by (Literal['groups', 'columns', 'columns_in_groups', 'cross'], optional): The method to compare the data. Defaults to 'groups'.
+            baseline_field (Optional[str], optional): The column to use as the baseline for comparison. Required if `compare_by` is 'columns' or 'columns_in_groups'. Defaults to None.
+
+        Returns:
+            Tuple: A tuple containing the baseline data and the compared data.
+
+        Raises:
+            NoRequiredArgumentError: If `baseline_field` is None and `compare_by` is 'columns' or 'columns_in_groups' or 'cross'.
+            ValueError: If `compare_by` is not one of the allowed values.
+        """
+        if compare_by == "groups":
+            if isinstance(target_fields, List):
+                target_fields = Adapter.list_to_single(target_fields)
+            if not isinstance(target_fields, str):
+                raise TypeError(
+                    f"group_field must be one string, {type(group_field)} passed."
+                )
+        elif baseline_field is None:
+            raise NoRequiredArgumentError("baseline_field")
+
+        if compare_by != "columns":
+            if group_field is None:
+                raise NoRequiredArgumentError("group_field")
+            elif isinstance(group_field, List):
+                group_field = Adapter.list_to_single(group_field)
+            if not isinstance(group_field, str):
+                raise TypeError(
+                    f"group_field must be one string, {type(group_field)} passed."
+                )
+
+        target_fields = Adapter.to_list(target_fields)
+        if compare_by == "groups":
+            data_buckets = data.groupby(by=group_field, fields_list=target_fields)
+            baseline_data = cls._split_ds_into_columns([data_buckets.pop(0)])
+            compared_data = cls._split_ds_into_columns(data=data_buckets)
+        elif compare_by == "columns":
+            baseline_data = [
+                (f"0{NAME_BORDER_SYMBOL}{baseline_field}", data[baseline_field])
+            ]
+            compared_data = [
+                (f"0{NAME_BORDER_SYMBOL}{column}", data[column])
+                for column in target_fields
+            ]
+        elif compare_by == "columns_in_groups":
+            baseline_data = cls._split_ds_into_columns(
+                data.groupby(by=group_field, fields_list=baseline_field)
+            )
+            compared_data = data.groupby(by=group_field, fields_list=target_fields)
+            compared_data = cls._split_ds_into_columns(data=compared_data)
+        elif compare_by == "cross":
+            data_buckets = data.groupby(by=group_field, fields_list=baseline_field)
+            baseline_data = cls._split_ds_into_columns([data_buckets.pop(0)])
+            compared_data = data.groupby(by=group_field, fields_list=target_fields)
+            compared_data.pop(0)
+            compared_data = cls._split_ds_into_columns(data=compared_data)
```

Comment:
may be u can do groupby outside if-else, and then do necessary actions

Reply:
It can be done but ill not make the code shorter, but will puzzle the logic split between scenarios.